crates/circuit/src/packed_instruction.rs

```diff
@@ -39,132 +42,375 @@ enum PackedOperationType {
+    }
+    const fn from_bits(value: u8) -> Self {
+        if value as usize >= STANDARD_GATE_SIZE {
+            panic!("unexpected standard gate type!")
+        }
+        // It would've been nice to use bytemuck here, but it doesn't support const
+        // casting for enums.
+        unsafe { std::mem::transmute(value) }
+    }
+}
+
+// Used by bitfield-struct to convert into and out of the bitfield.
+impl StandardInstructionType {
+    const fn into_bits(self) -> u8 {
+        self as _
+    }
+
+    const fn from_bits(value: u8) -> Self {
+        match value {
+            0 => StandardInstructionType::Barrier,
+            1 => StandardInstructionType::Delay,
+            2 => StandardInstructionType::Measure,
+            3 => StandardInstructionType::Reset,
+            _ => panic!("unexpected standard instruction type!"),
+        }
+    }
+}
+
+impl PackedOperationType {
+    const fn from_bits(value: u8) -> Self {
+        match value {
+            0 => Self::StandardGate,
+            1 => Self::StandardInstruction,
+            2 => Self::PyGatePointer,
+            3 => Self::PyInstructionPointer,
+            4 => Self::PyOperationPointer,
+            _ => panic!("unexpected discriminant type!"),
+        }
+    }
+}
+
+/// The bitfield layout used for standard gates.
+#[bitfield(u64)]
+struct StandardGateBits {
+    #[bits(3, default = PackedOperationType::StandardGate, access = RO)]
+    discriminant: PackedOperationType,
+    #[bits(8)]
+    standard_gate: StandardGate,
+    #[bits(53)]
+    __: u64,
+}
+
+/// The bitfield layout used for standard instructions.
+#[bitfield(u64)]
+struct StandardInstructionBits {
+    #[bits(3, default = PackedOperationType::StandardInstruction, access = RO)]
+    discriminant: PackedOperationType,
+    #[bits(8)]
+    standard_instruction: StandardInstructionType,
+    #[bits(21)]
+    _pad1: u32,
+    #[bits(32)]
+    value: ImmediateValue,
+}
+
+/// An inline value present within [StandardInstructionBits] layouts used to store arbitrary
+/// data to be interpreted according to the encoded standard instruction.
+#[derive(Clone, Copy, Debug)]
+#[repr(transparent)]
+struct ImmediateValue(u32);
+
+impl ImmediateValue {
+    const fn into_bits(self) -> u32 {
+        self.0
+    }
+
+    const fn from_bits(value: u32) -> Self {
+        Self(value)
+    }
+
+    #[inline]
+    fn from_delay_unit(unit: DelayUnit) -> Self {
+        Self(unit as u32)
+    }
+
+    #[inline]
+    fn delay_unit(&self) -> DelayUnit {
+        ::bytemuck::checked::cast(self.0 as u8)
+    }
+
+    #[inline]
+    fn u32(&self) -> u32 {
+        self.0
     }
+}
+
+/// The bitfield layout used by pointer types, like `PyGate`, `PyInstruction`, and `PyOperation`.
+///
+/// On 64-bit systems, the address is stored in the upper 62 bits, and is converted into and out of
+/// its packed form using `unpack_address` and `pack_address`. On 32-bit systems, the address is
+/// simply stored in the upper 32 bits.
```

Comment:
I think we could do "simply" better if we just do the same on both platforms, and have the `pointer` method cast the `u64` to `usize`.